File: app/src/main/java/com/lynn/simplerecyclerview/StyledEditText.java
package com.lynn.simplerecyclerview;

import android.content.Context;
import android.graphics.Typeface;
import android.support.v7.widget.AppCompatEditText;
import android.support.v7.widget.RecyclerView;
import android.text.Editable;
import android.text.Html;
import android.text.Spanned;
import android.text.style.StrikethroughSpan;
import android.text.style.StyleSpan;
import android.util.AttributeSet;

import org.apache.commons.lang.StringEscapeUtils;

/**
 * Created by Lynn.
 */

public class StyledEditText extends AppCompatEditText {
    public StyledEditText(Context context) {
        super(context);
    }

    public StyledEditText(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public StyledEditText(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    public void onBoldClick() {
        int selectionStart = getSelectionStart();
        int selectionEnd = getSelectionEnd();
        Editable text = getText();
        if (selectionStart < selectionEnd && selectionStart >= 0 && selectionEnd <= text.length()) {
            StyleSpan[] spans = text.getSpans(selectionStart, selectionEnd, StyleSpan.class);
            if (null != spans && spans.length > 0) {
                StyleSpan[] ss = text.getSpans(selectionStart, selectionStart + 1, StyleSpan.class);
                boolean isBoldExists = false;
                if (null != ss && ss.length > 0) {
                    for (int i = 0; i < ss.length; i++) {
                        if (ss[i].getStyle() == Typeface.BOLD) {
                            isBoldExists = true;
                        }
                    }
                }
                for (int i = 0; i < spans.length; i++) {
                    StyleSpan span = spans[i];
                    if (span.getStyle() == Typeface.BOLD) {
                        int spanStart = text.getSpanStart(span);
                        int spanEnd = text.getSpanEnd(span);
                        if (spanStart < selectionStart) {
                            text.setSpan(new StyleSpan(span.getStyle()), spanStart, selectionStart, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
                        }
                        if (spanEnd > selectionEnd) {
                            text.setSpan(new StyleSpan(span.getStyle()), selectionEnd, spanEnd, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
                        }
                        text.removeSpan(span);
                    }
                }
                if (!isBoldExists) {
                    text.setSpan(new StyleSpan(Typeface.BOLD), selectionStart, selectionEnd, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
                }
            } else {
                text.setSpan(new StyleSpan(Typeface.BOLD), selectionStart, selectionEnd, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
            }
        }
    }

    public void onItalicClick() {
        int selectionStart = getSelectionStart();
        int selectionEnd = getSelectionEnd();
        Editable text = getText();
        if (selectionStart < selectionEnd && selectionStart >= 0 && selectionEnd <= text.length()) {
            StyleSpan[] spans = text.getSpans(selectionStart, selectionEnd, StyleSpan.class);
            if (null != spans && spans.length > 0) {
                StyleSpan[] ss = text.getSpans(selectionStart, selectionStart + 1, StyleSpan.class);
                boolean isItalicExists = false;
                if (null != ss && ss.length > 0) {
                    for (int i = 0; i < ss.length; i++) {
                        if (ss[i].getStyle() == Typeface.ITALIC) {
                            isItalicExists = true;
                        }
                    }
                }
                for (int i = 0; i < spans.length; i++) {
                    StyleSpan span = spans[i];
                    if (span.getStyle() == Typeface.ITALIC) {
                        int spanStart = text.getSpanStart(span);
                        int spanEnd = text.getSpanEnd(span);
                        if (spanStart < selectionStart) {
                            text.setSpan(new StyleSpan(span.getStyle()), spanStart, selectionStart, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
                        }
                        if (spanEnd > selectionEnd) {
                            text.setSpan(new StyleSpan(span.getStyle()), selectionEnd, spanEnd, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
                        }
                        text.removeSpan(span);
                    }
                }
                if (!isItalicExists) {
                    text.setSpan(new StyleSpan(Typeface.ITALIC), selectionStart, selectionEnd, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
                }
            } else {
                text.setSpan(new StyleSpan(Typeface.ITALIC), selectionStart, selectionEnd, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
            }
        }
    }

    public void onStrikeClick() {
        int selectionStart = getSelectionStart();
        int selectionEnd = getSelectionEnd();
        Editable text = getText();
        if (selectionStart < selectionEnd && selectionStart >= 0 && selectionEnd <= text.length()) {
            StrikethroughSpan[] spans = text.getSpans(selectionStart, selectionEnd, StrikethroughSpan.class);
            if (null != spans && spans.length > 0) {
                StrikethroughSpan[] ss = text.getSpans(selectionStart, selectionStart + 1, StrikethroughSpan.class);
                boolean isItalicExists = false;
                if (null != ss && ss.length > 0) {
                    isItalicExists = true;
                }
                for (int i = 0; i < spans.length; i++) {
                    StrikethroughSpan span = spans[i];
                    int spanStart = text.getSpanStart(span);
                    int spanEnd = text.getSpanEnd(span);
                    if (spanStart < selectionStart) {
                        text.setSpan(new StrikethroughSpan(), spanStart, selectionStart, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
                    }
                    if (spanEnd > selectionEnd) {
                        text.setSpan(new StrikethroughSpan(), selectionEnd, spanEnd, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
                    }
                    text.removeSpan(span);
                }
                if (!isItalicExists) {
                    text.setSpan(new StrikethroughSpan(), selectionStart, selectionEnd, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
                }
            } else {
                text.setSpan(new StrikethroughSpan(), selectionStart, selectionEnd, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
            }
        }
    }

    public void onNormalClick() {
        int selectionStart = getSelectionStart();
        int selectionEnd = getSelectionEnd();
        Editable text = getText();
        if (selectionStart < selectionEnd && selectionStart >= 0 && selectionEnd <= text.length()) {
            StyleSpan[] spans = text.getSpans(selectionStart, selectionEnd, StyleSpan.class);
            if (null != spans && spans.length > 0) {
                for (int i = 0; i < spans.length; i++) {
                    StyleSpan span = spans[i];
                    int spanStart = text.getSpanStart(span);
                    int spanEnd = text.getSpanEnd(span);
                    if (spanStart < selectionStart) {
                        text.setSpan(new StyleSpan(span.getStyle()), spanStart, selectionStart, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
                    }
                    if (spanEnd > selectionEnd) {
                        text.setSpan(new StyleSpan(span.getStyle()), selectionEnd, spanEnd, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
                    }
                    text.removeSpan(span);
                }
            }
            StrikethroughSpan[] sspans = text.getSpans(selectionStart, selectionEnd, StrikethroughSpan.class);
            if (null != sspans && sspans.length > 0) {
                for (int i = 0; i < sspans.length; i++) {
                    StrikethroughSpan span = sspans[i];
                    int spanStart = text.getSpanStart(span);
                    int spanEnd = text.getSpanEnd(span);
                    if (spanStart < selectionStart) {
                        text.setSpan(new StrikethroughSpan(), spanStart, selectionStart, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
                    }
                    if (spanEnd > selectionEnd) {
                        text.setSpan(new StrikethroughSpan(), selectionEnd, spanEnd, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
                    }
                    text.removeSpan(span);
                }
            }
        }
    }

    public String toHtml() {
        Editable editable = getText();
        return StringEscapeUtils.unescapeHtml(Html.toHtml(editable));
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
        setEnabled(false);
        setEnabled(true);
    }
}

File: app/src/main/java/com/lynn/simplerecyclerview/common/StyledEditText.kt
package com.lynn.simplerecyclerview.common

import android.content.Context
import android.graphics.Typeface
import android.support.v7.widget.AppCompatEditText
import android.text.Html
import android.text.Spanned
import android.text.style.StrikethroughSpan
import android.text.style.StyleSpan
import android.util.AttributeSet

import org.apache.commons.lang.StringEscapeUtils

/**
 * Created by Lynn.
 */

class StyledEditText : AppCompatEditText {
    constructor(context : Context) : super(context) {}

    constructor(context : Context , attrs : AttributeSet) : super(context , attrs) {}

    constructor(context : Context , attrs : AttributeSet , defStyleAttr : Int) : super(context , attrs , defStyleAttr) {}

    fun onBoldClick() {
        val selectionStart = selectionStart
        val selectionEnd = selectionEnd
        val text = text
        if (selectionStart in 0..(selectionEnd - 1) && selectionEnd <= text.length) {
            val spans = text.getSpans(selectionStart , selectionEnd , StyleSpan::class.java)
            if (null != spans && spans.isNotEmpty()) {
                val ss = text.getSpans(selectionStart , selectionStart + 1 , StyleSpan::class.java)
                var isBoldExists = false
                if (null != ss && ss.isNotEmpty()) {
                    ss.indices
                            .filter { ss[it].style == Typeface.BOLD }
                            .forEach { isBoldExists = true }
                }
                for (i in spans.indices) {
                    val span = spans[i]
                    if (span.style == Typeface.BOLD) {
                        val spanStart = text.getSpanStart(span)
                        val spanEnd = text.getSpanEnd(span)
                        if (spanStart < selectionStart) {
                            text.setSpan(StyleSpan(span.style) , spanStart , selectionStart , Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
                        }
                        if (spanEnd > selectionEnd) {
                            text.setSpan(StyleSpan(span.style) , selectionEnd , spanEnd , Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
                        }
                        text.removeSpan(span)
                    }
                }
                if (!isBoldExists) {
                    text.setSpan(StyleSpan(Typeface.BOLD) , selectionStart , selectionEnd , Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
                }
            } else {
                text.setSpan(StyleSpan(Typeface.BOLD) , selectionStart , selectionEnd , Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
            }
        }
    }

    fun onItalicClick() {
        val selectionStart = selectionStart
        val selectionEnd = selectionEnd
        val text = text
        if (selectionStart in 0..(selectionEnd - 1) && selectionEnd <= text.length) {
            val spans = text.getSpans(selectionStart , selectionEnd , StyleSpan::class.java)
            if (null != spans && spans.isNotEmpty()) {
                val ss = text.getSpans(selectionStart , selectionStart + 1 , StyleSpan::class.java)
                var isItalicExists = false
                if (null != ss && ss.isNotEmpty()) {
                    ss.indices
                            .filter { ss[it].style == Typeface.ITALIC }
                            .forEach { isItalicExists = true }
                }
                for (i in spans.indices) {
                    val span = spans[i]
                    if (span.style == Typeface.ITALIC) {
                        val spanStart = text.getSpanStart(span)
                        val spanEnd = text.getSpanEnd(span)
                        if (spanStart < selectionStart) {
                            text.setSpan(StyleSpan(span.style) , spanStart , selectionStart , Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
                        }
                        if (spanEnd > selectionEnd) {
                            text.setSpan(StyleSpan(span.style) , selectionEnd , spanEnd , Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
                        }
                        text.removeSpan(span)
                    }
                }
                if (!isItalicExists) {
                    text.setSpan(StyleSpan(Typeface.ITALIC) , selectionStart , selectionEnd , Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
                }
            } else {
                text.setSpan(StyleSpan(Typeface.ITALIC) , selectionStart , selectionEnd , Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
            }
        }
    }

    fun onStrikeClick() {
        val selectionStart = selectionStart
        val selectionEnd = selectionEnd
        val text = text
        if (selectionStart in 0..(selectionEnd - 1) && selectionEnd <= text.length) {
            val spans = text.getSpans(selectionStart , selectionEnd , StrikethroughSpan::class.java)
            if (null != spans && spans.isNotEmpty()) {
                val ss = text.getSpans(selectionStart , selectionStart + 1 , StrikethroughSpan::class.java)
                var isItalicExists = false
                if (null != ss && ss.isNotEmpty()) {
                    isItalicExists = true
                }
                for (i in spans.indices) {
                    val span = spans[i]
                    val spanStart = text.getSpanStart(span)
                    val spanEnd = text.getSpanEnd(span)
                    if (spanStart < selectionStart) {
                        text.setSpan(StrikethroughSpan() , spanStart , selectionStart , Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
                    }
                    if (spanEnd > selectionEnd) {
                        text.setSpan(StrikethroughSpan() , selectionEnd , spanEnd , Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
                    }
                    text.removeSpan(span)
                }
                if (!isItalicExists) {
                    text.setSpan(StrikethroughSpan() , selectionStart , selectionEnd , Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
                }
            } else {
                text.setSpan(StrikethroughSpan() , selectionStart , selectionEnd , Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
            }
        }
    }

    fun onNormalClick() {
        val selectionStart = selectionStart
        val selectionEnd = selectionEnd
        val text = text
        if (selectionStart in 0..(selectionEnd - 1) && selectionEnd <= text.length) {
            val spans = text.getSpans(selectionStart , selectionEnd , StyleSpan::class.java)
            if (null != spans && spans.isNotEmpty()) {
                for (i in spans.indices) {
                    val span = spans[i]
                    val spanStart = text.getSpanStart(span)
                    val spanEnd = text.getSpanEnd(span)
                    if (spanStart < selectionStart) {
                        text.setSpan(StyleSpan(span.style) , spanStart , selectionStart , Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
                    }
                    if (spanEnd > selectionEnd) {
                        text.setSpan(StyleSpan(span.style) , selectionEnd , spanEnd , Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
                    }
                    text.removeSpan(span)
                }
            }
            val sspans = text.getSpans(selectionStart , selectionEnd , StrikethroughSpan::class.java)
            if (null != sspans && sspans.isNotEmpty()) {
                for (i in sspans.indices) {
                    val span = sspans[i]
                    val spanStart = text.getSpanStart(span)
                    val spanEnd = text.getSpanEnd(span)
                    if (spanStart < selectionStart) {
                        text.setSpan(StrikethroughSpan() , spanStart , selectionStart , Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
                    }
                    if (spanEnd > selectionEnd) {
                        text.setSpan(StrikethroughSpan() , selectionEnd , spanEnd , Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
                    }
                    text.removeSpan(span)
                }
            }
        }
    }

    fun toHtml() : String? {
        val editable = text
        return StringEscapeUtils.unescapeHtml(Html.toHtml(editable))
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        isEnabled = false
        isEnabled = true
    }
}
